File: skeleton/src/me/shkschneider/skeleton/helpers/DateTimeHelper.java
package me.shkschneider.skeleton.helpers;

import android.text.format.DateUtils;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.TimeZone;

public class DateTimeHelper {

    public static Calendar calendar(final int year, final int month, final int day, final int hour, final int minutes, final int seconds) {
        final Calendar calendar = Calendar.getInstance();
        calendar.set(Calendar.DAY_OF_MONTH, day);
        calendar.set(Calendar.MONTH, month);
        calendar.set(Calendar.YEAR, year);
        calendar.set(Calendar.HOUR_OF_DAY, hour);
        calendar.set(Calendar.MINUTE, minutes);
        calendar.set(Calendar.SECOND, seconds);
        return calendar;
    }

    public static Calendar calendar(final int year, final int month, final int day) {
        return calendar(day, month, year, 0, 0, 0);
    }

    public static int year() {
        return Calendar.getInstance().get(Calendar.YEAR);
    }

    public static int month() {
        return Calendar.getInstance().get(Calendar.MONTH) + 1;
    }

    public static int day() {
        return Calendar.getInstance().get(Calendar.DAY_OF_MONTH);
    }

    public static String format(final String format, final Calendar calendar) {
        final SimpleDateFormat simpleDateFormat = new SimpleDateFormat(format);
        return simpleDateFormat.format(calendar.getTime());
    }

    public static long timestamp() {
        return Calendar.getInstance().getTimeInMillis() / 1000;
    }

    public static long millitimestamp() {
        return Calendar.getInstance().getTimeInMillis();
    }

    private static int gmtOffset() {
        return Math.round(TimeZone.getDefault().getOffset(millitimestamp()) / DateUtils.SECOND_IN_MILLIS);
    }

    public static String relative(final long time) {
        if (time < 0) {
            LogHelper.warning("Time was invalid");
            return null;
        }

        return DateUtils.getRelativeTimeSpanString(time, timestamp(), DateUtils.SECOND_IN_MILLIS, DateUtils.FORMAT_ABBREV_RELATIVE).toString();
    }

    public static String relative(final long from, final long to) {
        if (from < 0) {
            LogHelper.warning("TimeFrom was invalid");
            return null;
        }
        if (to < 0) {
            LogHelper.warning("TimeTo was invalid");
            return null;
        }

        return DateUtils.getRelativeTimeSpanString(from, to, DateUtils.SECOND_IN_MILLIS, DateUtils.FORMAT_ABBREV_RELATIVE).toString();
    }

}

File: android/src/main/kotlin/me/shkschneider/skeleton/android/util/_DateTimeHelper.kt
package me.shkschneider.skeleton.android.util

import android.app.Activity
import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.text.format.DateFormat
import android.text.format.DateUtils
import androidx.annotation.IntRange
import me.shkschneider.skeleton.android.os.AndroidHelper
import me.shkschneider.skeleton.kotlin.util.DateTimeHelper
import java.util.Calendar

val DateTimeHelper.RFC_3339
    get() =
        if (AndroidHelper.api() >= AndroidHelper.ANDROID_7) "yyyy-MM-dd'T'HH:mm:ssXXX"
        else "yyyy-MM-dd'T'HH:mm:ssZ"

fun DateTimeHelper.relative(@IntRange(from = 0) time: Long): String =
    relative(time, time)

fun DateTimeHelper.relative(@IntRange(from = 0) from: Long, @IntRange(from = 0) to: Long): String =
    DateUtils.getRelativeTimeSpanString(from, to, DateUtils.SECOND_IN_MILLIS, DateUtils.FORMAT_ABBREV_RELATIVE).toString()

fun DateTimeHelper.pickTime(activity: Activity, date: Calendar? = null, onTimeSetListener: TimePickerDialog.OnTimeSetListener) {
    val calendar = date ?: Calendar.getInstance()
    val is24HourFormat = DateFormat.is24HourFormat(activity)
    TimePickerDialog(activity, onTimeSetListener, calendar.get(Calendar.HOUR_OF_DAY), calendar.get(Calendar.MINUTE), is24HourFormat).show()
}

fun DateTimeHelper.pickDate(activity: Activity, min: Calendar? = null, date: Calendar? = null, max: Calendar? = null, onDateSetListener: DatePickerDialog.OnDateSetListener) {
    val calendar = date ?: Calendar.getInstance() // now by default
    val datePickerDialog = DatePickerDialog(activity,
        onDateSetListener,
        calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH))
    val datePicker = datePickerDialog.datePicker
    min?.let {
        datePicker.minDate = 0 // HACK: <http://stackoverflow.com/a/19722636>
        datePicker.minDate = it.timeInMillis
    }
    max?.let {
        datePicker.maxDate = it.timeInMillis
    }
    datePickerDialog.show()
}
